tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py

```diff
@@ -239,15 +244,206 @@ async def with_ci_connector_ops(context: PipelineContext) -> Container:
     """
     python_base_environment: Container = with_python_base(context, "python:3-alpine")
     python_with_git = with_alpine_packages(python_base_environment, ["gcc", "libffi-dev", "musl-dev", "git"])
-    return await with_installed_local_python_package(context, python_with_git, CI_CONNECTOR_OPS_SOURCE_PATH, exclude=["pipelines"])
+    return await with_installed_python_package(context, python_with_git, CI_CONNECTOR_OPS_SOURCE_PATH, exclude=["pipelines"])
+
+
+def with_dockerd_service(
+    context: ConnectorTestContext, shared_volume: Optional(Tuple[str, CacheVolume]) = None, docker_service_name: Optional[str] = None
+) -> Container:
+    """Create a container running dockerd, exposing its 2375 port, can be used as the docker host for docker-in-docker use cases.
+
+    Args:
+        context (ConnectorTestContext): The current connector test context.
+        shared_volume (Optional, optional): A tuple in the form of (mounted path, cache volume) that will be mounted to the dockerd container. Defaults to None.
+        docker_service_name (Optional[str], optional): The name of the docker service, appended to volume name, useful context isolation. Defaults to None.
+
+    Returns:
+        Container: The container running dockerd as a service.
+    """
+    docker_lib_volume_name = f"{slugify(context.connector.technical_name)}-docker-lib"
+    if docker_service_name:
+        docker_lib_volume_name = f"{docker_lib_volume_name}-{slugify(docker_service_name)}"
+    dind = (
+        context.dagger_client.container()
+        .from_("docker:23.0.1-dind")
+        .with_mounted_cache(
+            "/var/lib/docker",
+            context.dagger_client.cache_volume(docker_lib_volume_name),
+            sharing=CacheSharingMode.SHARED,
+        )
+    )
+    if shared_volume is not None:
+        dind = dind.with_mounted_cache(*shared_volume)
+    return dind.with_exposed_port(2375).with_exec(
+        ["dockerd", "--log-level=error", "--host=tcp://0.0.0.0:2375", "--tls=false"], insecure_root_capabilities=True
+    )
+
+
+def with_bound_docker_host(
```

Comment:
Well done with all of this. This type of code is not easy at all. Kudos.

Reply:
Ahah this is indeed a good portion of the complexity of this PR